packages/core/src/extensions/Blocks/nodes/BlockContainer.ts

```diff
@@ -192,18 +192,42 @@ export const BlockContainer = Node.create<{
               );
             }
 
-            // Changes the blockContent node type and adds the provided props as attributes. Also preserves all existing
-            // attributes that are compatible with the new type.
-            state.tr.setNodeMarkup(
-              startPos,
-              block.type === undefined
-                ? undefined
-                : state.schema.nodes[block.type],
-              {
-                ...contentNode.attrs,
-                ...block.props,
-              }
-            );
+            // Since some block types contain inline content and others don't,
+            // we either need to call setNodeMarkup to just update type &
+            // attributes, or replaceWith to replace the whole blockContent.
+            const oldType = contentNode.type.name;
+            const newType = block.type || oldType;
+
+            const oldContentType = state.schema.nodes[oldType].spec.content;
+            const newContentType = state.schema.nodes[newType].spec.content;
+
+            if (oldContentType === "inline*" && newContentType === "") {
+              // Replaces the blockContent node with one of the new type and
+              // adds the provided props as attributes. Also preserves all
+              // existing attributes that are compatible with the new type.
+              state.tr.replaceWith(
```

Comment:
can't we always use replaceWith to simplify this?

And what if the new one has Inline, but the old doesn't?

Reply:
If the old one doesn't have inline and the new one does, then it's no problem since PM knows that it just needs to fill in the empty content with a paragraph node. It's only a problem the other way around since you are trying to add content to a node that doesn't accept any content.